processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

```diff
@@ -51,25 +63,61 @@ public SelectorDimFilter(
   {
     byte[] dimensionBytes = StringUtils.toUtf8(dimension);
     byte[] valueBytes = (value == null) ? new byte[]{} : StringUtils.toUtf8(value);
+    byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
 
-    return ByteBuffer.allocate(2 + dimensionBytes.length + valueBytes.length)
+    return ByteBuffer.allocate(3 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
                      .put(DimFilterCacheHelper.SELECTOR_CACHE_ID)
                      .put(dimensionBytes)
                      .put(DimFilterCacheHelper.STRING_SEPARATOR)
                      .put(valueBytes)
+                     .put(DimFilterCacheHelper.STRING_SEPARATOR)
+                     .put(extractionFnBytes)
                      .array();
   }
 
   @Override
   public DimFilter optimize()
   {
+    if (this.getExtractionFn() instanceof LookupExtractionFn
+        && ((LookupExtractionFn) this.getExtractionFn()).isOptimize()) {
+      LookupExtractionFn exFn = (LookupExtractionFn) this.getExtractionFn();
+      LookupExtractor lookup = exFn.getLookup();
+
+      final String convertedValue = Strings.emptyToNull(value);
+
+      // We cannot do an unapply()-based optimization if the selector value
+      // and the replaceMissingValuesWith value are the same, since we have to match on
+      // all values that are not present in the lookup.
+      if (!exFn.isRetainMissingValue() && Objects.equals(convertedValue, exFn.getReplaceMissingValueWith())) {
+        return this;
+      }
+
+      final String mappingForValue = lookup.apply(convertedValue);
+      final List<String> keys = lookup.unapply(convertedValue);
+
+      // If retainMissingValues is true and the selector value is not in the lookup map,
+      // there may be row values that match the selector value but are not included
+      // in the lookup map. Match on the selector value as well.
+      // If the selector value is overwritten in the lookup map, don't add selector value to keys.
+      if (exFn.isRetainMissingValue() && mappingForValue == null) {
+        keys.add(convertedValue);
```

Comment:
`keys` isn't guaranteed to be mutable, so it'd be better to copy it and add `convertedValue` to the copy.